src/renderer/CharacterDisplayRenderer.h

```diff
@@ -0,0 +1,120 @@
+#pragma once
+
+#include "MenuRenderer.h"
+#include "display/CharacterDisplayInterface.h"
+
+/**
+ * @class CharacterDisplayRenderer
+ * @brief A renderer for character-based displays, extending the MenuRenderer class.
+ *
+ * This class provides functionality to render menus on character-based displays.
+ * It supports custom icons for cursor and edit cursor, as well as up and down arrows
+ * for navigation indicators.
+ *
+ * @details
+ * The CharacterDisplayRenderer class is designed to work with displays that implement
+ * the CharacterDisplayInterface. It provides methods to draw menu items, cursors, and
+ * navigation indicators. The class also handles text truncation and cursor movement.
+ *
+ * @note
+ * The class uses dynamic memory allocation for the upArrow and downArrow icons.
+ */
+class CharacterDisplayRenderer : public MenuRenderer {
+  protected:
+    uint8_t* upArrow;
+    uint8_t* downArrow;
+    const uint8_t cursorIcon;
+    const uint8_t editCursorIcon;
```

Comment:
I think `MenuScreen` can be responsible for this. 
Particular implementation of `MenuScreen` can register custom chars and control which and where to draw. Also it can control the layout, need just change method signature for `MenuItem` from `DisplayInterface* display, uint8_t row` to `DisplayInterface* display, uint8_t row, uint8_t  col, uint8_t  size)` to tell menu item where it starts and where it should ends. Then different implementations of `MenuScreen` can decide, for example current impl draws from `1 to n-2` and keep the first and last cols for cursors, another impl can draws `from 0 to n`

Reply:
See this https://lcdmenu--233.org.readthedocs.build/en/233/overview/rendering/character-display.html#customizing-the-character-display-renderer.

If I'm not mistaken, this is handled already.

And devs can also provide custom renderers (if they don't want to display the cursor in pos `1`).

Having the renderer, I think it's easier to customize rendering on the screen, you just need to override the `drawItem` function and say how your item should be drawn, `drawItem` receives the text like `TEXT:VALUE`, then it adds `>TEXT:VALUE    ↓` and renders, you can change easily change that. To customize the semicolon, would be better if you create a custom item type.

Reply:
All in 1 place:
```c++
void MenuScreen::init(DisplayInterface* display) const {
    // Register chars
    display.createChar(0, UP_ARROW);
    display.createChar(1, DOWN_ARROW);
}

void MenuScreen::draw(DisplayInterface* display) const {
    // Draw items
    bool notFullView = false;
    for (uint8_t i = 0; i < display->getMaxRows(); i++) {
        MenuItem* item = this->items[view + i];
        if (item == nullptr) {
            notFullView = true;
            break;
        }
        item->draw(display, i, 1, display->getMaxRows() - 2); // <- Skip first and last cols.
    }
    // Draw scrolling
    display->setCursor(0, 0);
    if (view == 0) {
        display->draw(' ');
    } else {
        display->draw('\000');
    }
    display->setCursor(getMaxCols() - 1, getMaxRows() - 1);
    if (notFullView || items[view + display->getMaxRows()] == nullptr) {
        display->draw(' ');
    } else {
        display->draw("\001");
    }
    // Draw cursor
    display->setCursor(0, cursor - view);
    display->draw(editMode ? 0x7E : 0x7F);
}
```

Another impl can have:

```c++
void MenuScreen::draw(DisplayInterface* display) const {
    // Draw items
    for (uint8_t i = 0; i < display->getMaxRows(); i++) {
        MenuItem* item = this->items[view + i];
        if (item == nullptr) {
            return;
        }
        item->draw(display, i, 0, display->getMaxRows() - 1); // <- Do not skip cols, draw items on all width
    }
}
```

Reply:
It's just about responsibility, not about lack of functionality. 
Right now currently active item is tracked in menu screen and also is tracked inside menu renderer, `itemCount` also..

Reply:
I think having all that in `MenuScreen` or subclassing `MenuScreen` reduces flexibility, in the future when other display types are supported, it'll be easier to change how items are rendered to match your needs, especially for displays that allow much customization like TFT displays, as you can see in this impl, `MenuScreen` knows nothing about `cursorIcons` etc, it's the responsibility of the renderer. In the case of TFT display, cursor is a different concept and we would then define how we render a cursor (it might be just surround the row with a box, or invert the bg and fg color etc.)

It also makes it easier to swap renderers and displays easily without needing to redefine menu screen.

Reply:
Your example with different concept of cursor convinced me 😄

Reply:
Can it be named `LcdRenderer` then? then will be added `TftRenderer`..

Reply:
I think `CharacterRenderer` is fitting, since this type of display is character display. There are other LCD's that are not character based displays.